apps/prairielearn/src/lib/preact.tsx

```diff
@@ -0,0 +1,70 @@
+/* eslint-disable @typescript-eslint/no-unused-vars */
+import { render } from 'preact-render-to-string/jsx';
+
+import { compiledScriptPath } from '@prairielearn/compiled-assets';
+import { type HtmlSafeString } from '@prairielearn/html';
+import {
+  type Attributes,
+  type ComponentType,
+  Fragment,
+  type VNode,
+} from '@prairielearn/preact-cjs';
+
+import { renderHtml } from './preact-html.js';
+
+// Based on https://pkg.go.dev/encoding/json#HTMLEscape
+const ENCODE_HTML_RULES: Record<string, string> = {
+  '&': '\\u0026',
+  '>': '\\u003e',
+  '<': '\\u003c',
+  '\u2028': '\\u2028',
+  '\u2029': '\\u2029',
+};
+const MATCH_HTML = /[&><\u2028\u2029]/g;
+
+function escapeJsonForHtml(value: any): string {
+  return JSON.stringify(value).replace(MATCH_HTML, (c) => ENCODE_HTML_RULES[c] || c);
+}
+
+export function renderHtmlDocument(content: VNode) {
+  return `<!doctype html>\n${render(content, {}, { pretty: true, jsx: false })}`;
+}
+
+export function renderForClientHydration<T>(
```

Comment:
Nit: now that we're actually landing this stuff, some JSDoc comments could be nice.